src/awkward/operations/ak_with_field.py

```diff
@@ -109,22 +110,21 @@ def action_is_record(input, **kwargs):
         if not purelist_is_record(base):
             raise ValueError("no tuples or records in array; cannot add a new field")
 
-        def action(inputs, **kwargs):
+        def action(inputs, depth_context, **kwargs):
             base, what = inputs
-            backend = base.backend
 
             if isinstance(base, ak.contents.RecordArray):
-                if what is None:
-                    what = ak.contents.IndexedOptionArray(
-                        ak.index.Index64(
-                            backend.index_nplike.full(len(base), -1, dtype=np.int64),
-                            nplike=backend.index_nplike,
-                        ),
-                        ak.contents.EmptyArray(),
-                    )
-                elif not isinstance(what, ak.contents.Content):
-                    what = ak.contents.NumpyArray(
-                        backend.nplike.repeat(what, len(base))
+                # Broadcast `what` with the first content (to the appropriate length)
+                if len(base.contents) > 0:
+                    first_content = base.contents[0]
+                    what, _ = broadcast_arrays(
+                        [what, first_content[: base.length]],
+                        depth_limit=1,
+                        left_broadcast=True,
+                        right_broadcast=True,
+                        broadcast_parameters_rule="one_to_one",
+                        highlevel=False,
+                        behavior=behavior,
                     )
```

Comment:
It feels like there should be a more elegant way to recurse _into_ the record array, but  I think this is actually the simplest solution.